src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -650,6 +662,141 @@ vertexSet = Set.fromAscList . vertexList
 edgeSet :: (Ord a, Ord b) => AdjacencyMap a b -> Set.Set (a, b)
 edgeSet = Set.fromAscList . edgeList
 
+-- | The /circuit/ on a list of vertices.
+-- Complexity: /O(n * log(n))/ time and /O(n)/ memory.
+--
+-- @
+-- circuit []                       == 'empty'
+-- circuit [(x, y)]                 == 'edge' x y
+-- circuit [(x, y), (z, w)]         == 'connect' ('vertices' [x, z] []) ('vertices' [] [y, w])
+-- circuit [(1, 2), (3, 4), (5, 6)] == swap 1 * (2 + 6) + swap 3 * (2 + 4) + swap 5 * (6 + 2)
+-- circuit . 'reverse'              == 'swap' . circuit . 'map' 'Data.Tuple.swap'
+-- @
+circuit :: (Ord a, Ord b) => [(a, b)] -> AdjacencyMap a b
+circuit l = overlay (edges l) (edges (zipWith (,) (map fst l) (rotateSnd l)))
+    where
+        rotateSnd :: [(a, b)] -> [b]
+        rotateSnd l = (snd (last l)):(map snd (init l))
+
+-- | The /biclique/ on two lists of vertices.
+-- Complexity: /O(n * log(n) + m)/ time and /O(n + m)/ memory.
+--
+-- @
+-- biclique [] [] == 'empty'
+-- biclique xs [] == 'vertices' xs []
+-- biclique [] ys == 'vertices' [] ys
+-- biclique xs ys == 'connect' ('vertices' xs []) ('vertices' [] ys)
+-- @
+biclique :: (Ord a, Ord b) => [a] -> [b] -> AdjacencyMap a b
+biclique xs ys = let sxs = sort xs
+                     sys = sort ys
+                  in BAM (Map.fromAscList $ zipWith (,) sxs $ repeat (Set.fromAscList sys))
+                         (Map.fromAscList $ zipWith (,) sys $ repeat (Set.fromAscList sxs))
+
+data Part = LeftPart | RightPart
+    deriving (Show, Eq)
+
+otherPart :: Part -> Part
+otherPart LeftPart  = RightPart
+otherPart RightPart = LeftPart
+
+type PartMap a = Map.Map a Part
+type PartMonad a = MaybeT (State (PartMap a)) [a]
+
+-- | An odd cycle. For example, @[1, 2, 3]@ represents the cycle 1 → 2 → 3 → 1.
+type OddCycle a = [a] -- TODO: Make this representation type-safe
+
+neighbours :: Ord a => a -> AM.AdjacencyMap a -> [a]
+neighbours v = Set.toAscList . AM.postSet v
+
+partMonad :: Ord a => AM.AdjacencyMap a -> PartMonad a
+partMonad g = asum $ map (runDfs g) $ AM.vertexList g
```

Comment:
It's a minor tweak, but I often find list comprehensions easier to understand than layers of `map`s.

```suggestion
partMonad g = asum [ runDfs g v | v <- AM.vertexList g ]
```